contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,299 @@
+
+        // NftCollection logic
+        // check if the deal is nft gated
+        // if yes, store it in `nftCollectionDetails` and `nftId` and emit respective events for 721 and 1155
+        IAelinPool.NftCollectionRules[] memory nftCollectionRules = _poolData.nftCollectionRules;
+
+        if (nftCollectionRules.length > 0) {
+            // if the first address supports punks or 721, the entire pool only supports 721 or punks
+            if (
+                nftCollectionRules[0].collectionAddress == CRYPTO_PUNKS ||
+                NftCheck.supports721(nftCollectionRules[0].collectionAddress)
+            ) {
+                for (uint256 i = 0; i < nftCollectionRules.length; i++) {
+                    require(
+                        nftCollectionRules[i].collectionAddress == CRYPTO_PUNKS ||
+                            NftCheck.supports721(nftCollectionRules[i].collectionAddress),
+                        "can only contain 721"
+                    );
+                    nftCollectionDetails[nftCollectionRules[i].collectionAddress] = nftCollectionRules[i];
+                    emit PoolWith721(
+                        nftCollectionRules[i].collectionAddress,
+                        nftCollectionRules[i].purchaseAmount,
+                        nftCollectionRules[i].purchaseAmountPerToken
+                    );
+                }
+                hasNftList = true;
+            }
+            // if the first address supports 1155, the entire pool only supports 1155
+            else if (NftCheck.supports1155(nftCollectionRules[0].collectionAddress)) {
+                for (uint256 i = 0; i < nftCollectionRules.length; i++) {
+                    require(NftCheck.supports1155(nftCollectionRules[i].collectionAddress), "can only contain 1155");
+                    nftCollectionDetails[nftCollectionRules[i].collectionAddress] = nftCollectionRules[i];
+
+                    for (uint256 j = 0; j < nftCollectionRules[i].tokenIds.length; j++) {
+                        nftId[nftCollectionRules[i].collectionAddress][nftCollectionRules[i].tokenIds[j]] = true;
+                    }
+                    emit PoolWith1155(
+                        nftCollectionRules[i].collectionAddress,
+                        nftCollectionRules[i].purchaseAmount,
+                        nftCollectionRules[i].purchaseAmountPerToken,
+                        nftCollectionRules[i].tokenIds,
+                        nftCollectionRules[i].minTokensEligible
+                    );
+                }
+                hasNftList = true;
+            } else {
+                revert("collection is not compatible");
+            }
+        }
+
+        // deposit underlying token logic
+        // check if the underlyingDealAmount is more than 0,
+        // if yes, transfer it to this contract and store it in `currentDealTokenTotal` and add it
+        if (_underlyingDealTokenAmount > 0) {
+            currentDealTokenTotal += _underlyingDealTokenAmount;
```

Comment:
this is potentially a bug. you are giving them credit for the full amount of the transfer but it is possible that some tokens will burn some of the supply on a transfer. in this case we need something like this to protect ourselves. also, the transfer is coming from the holder not the sponsor. we might need to clarify the roles here I suspect 

IERC20(_dealData.underlyingDealToken).transferFrom(_sponsor, address(this), _underlyingDealTokenAmount);
uint256 balanceAfter =  IERC20(_dealData.underlyingDealToken).balanceOf(address(this));

if (_underlyingDealTokenAmount >= balanceAfter) ...

if (


IERC20(_dealData.underlyingDealToken).transferFrom(_sponsor, address(this), _underlyingDealTokenAmount);